util/collate/charset.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collate
+
+import (
+	"strings"
+
+	"github.com/pingcap/parser/charset"
+)
+
+var (
+	enableCharsetFeat bool
+)
+
+// EnableCharsetFeat the charset feature
+func EnableCharsetFeat() {
+	enableCharsetFeat = true
+	addCharset()
+}
+
+// SetCharsetFratEnabledForTest set charset feature enabled. Only used in test.
+// It will also enable or disable new collation.
+func SetCharsetFratEnabledForTest(flag bool) {
+	enableCharsetFeat = flag
+	SetNewCollationEnabledForTest(flag)
+	if flag {
+		addCharset()
+	} else {
+		removeCharset()
+	}
+}
+
+// CharsetFeatEnabled return true if charset feature is enabled.
+func CharsetFeatEnabled() bool {
+	return enableCharsetFeat
+}
+
+func addCharset() {
+	for _, c := range experimentalCharsetInfo {
+		charset.AddCharset(c)
+		for _, coll := range charset.GetCollations() {
+			if strings.EqualFold(coll.CharsetName, c.Name) {
+				charset.AddCollation(coll)
```

Comment:
Could we remove this code？
`GetCollations` gets collations from `charset.collations` and `AddCollation` adds a collation to `charset.collations`.

Reply:
nope, `AddCollation` not only add it to `charset.collations`, it also add collation to `Charset.Collations` field.

Reply:
Then, do we need to remove it when we do `removeCharset`?

Reply:
But we have done `AddCollation` in `init` for all `charset.collations`(include gbk).

Reply:
if charset is removed, it is removed.

Reply:
It seems that `charset.supportedCollations` doesn't remove it, when we do `show collation`(after `removeCharset`), the result is not what we expected.

Reply:
`supportedCollations` is not changed in `AddCollation`, when `AddCollation` only the collation in `supportedCollationNames` will be added into `supportedCollations`.

Reply:
Got it.